src/cargo/util/toml/targets.rs

```diff
@@ -1037,9 +1057,19 @@ pub fn normalize_build(build: Option<&StringOrBool>, package_root: &Path) -> Opt
             }
         }
         // Explicitly no build script.
-        Some(StringOrBool::Bool(false)) | Some(StringOrBool::String(_)) => build.cloned(),
+        Some(StringOrBool::Bool(false)) => build.cloned(),
+        Some(StringOrBool::String(build_file)) => {
+            let build_file = paths::normalize_path(Path::new(build_file));
+            let build = build_file
+                .into_os_string()
+                .into_string()
+                .map_err(|_err| anyhow::format_err!("non-UTF8 `package.build`"))?;
+            Some(StringOrBool::String(build))
```

Comment:
The error message is inaccurate as `package.build` can only be UTF-8 because `StringOrBool::String` is a `String`.

This could instead be
```rust
.expect("`build_file` started as a String and `normalize_path` shouldn't have changed that")
```